File: src/main/java/tooko/pm/handlers/EditHandler.java
package tooko.pm.handlers;

import tooko.main.*;
import tooko.main.bots.*;
import tooko.pm.*;
import tooko.td.TdApi.*;
import tooko.td.client.*;

public class EditHandler extends TdHandler {

    public PmBot pm;

    public BotData bot;
    public PmData data;

    public Lang OL;

    @Override
    public void onLoad() {

        pm = (PmBot) client;

        bot = pm.bot;
        data = pm.data;

        OL = pm.OL;

    }

    @Override
    public void onMessageContent(long chatId, long messageId, MessageContent newContent) {

        if (chatId == bot.owner) {

            String messageIdStr = messageId + "";

            if (data.sended.containsKey(messageIdStr)) {

                long targetMessage = data.sended.get(messageIdStr);

                PmData.Session session = data.getSessionByElement("sended", targetMessage);

                long targetChat = session.chatId;

                Message oldMessage = message(targetChat, targetMessage);

                if (oldMessage == null) {

                    data.sended.remove(messageIdStr);

                    session.sended.remove(targetMessage);

                    return;

                }

                Function edit = null;

                if (newContent instanceof MessageText) {

                    edit = Fn.editText(targetChat, targetMessage, ((MessageText) newContent).text);

                } else if (newContent instanceof MessageAnimation || newContent instanceof MessageAudio || newContent instanceof MessageDocument || newContent instanceof MessagePhoto || newContent instanceof MessageVideo) {

                    edit = new EditMessageMedia(targetChat, targetMessage, null, Fn.convertToInput(newContent));

                } else if (newContent instanceof MessageVoiceNote) {

                    edit = new EditMessageCaption(targetChat, targetMessage, null, ((MessageVoiceNote) newContent).caption);

                } else if (newContent instanceof MessageLocation) {

                    edit = new EditMessageLiveLocation(targetChat, targetMessage, null, ((MessageLocation) newContent).location);

                }

                if (edit == null) {

                    send(Fn.sendText(chatId, Fn.plainText(OL.EDIT_TYPE_INVALID)));

                    return;

                }

                try {

                    execute(edit);

                    delay(Fn.sendText(chatId, messageId, Fn.plainText(OL.EDIT_SUCCEED)));

                } catch (TdException e) {

                    send(Fn.sendText(chatId, messageId, Fn.plainText(OL.EDIT_FAILED, e)));

                }

            }

        } else {

            if (data.received.containsValue(messageId)) {

                long receivedMessageId = Long.parseLong(data.received.getKey(messageId));

                Message reportMessage = syncE(Fn.sendText(bot.owner, receivedMessageId, Fn.plainText(OL.EDITED_BY)));
                Message newMessage = syncE(Fn.sendMessage(bot.owner, Fn.forward(chatId, messageId)));

                data.sessions.arrayInsert(chatId, "extras", reportMessage.id);
                data.sessions.arrayInsert(chatId, "extras", newMessage.id);

            }

        }

    }


	/*

	 private boolean formatedTextEquals(FormattedText text,FormattedText newText) {

	 if (!text.text.equals(newText.text)) return false;

	 if (text.entities.length != newText.entities.length) return false;

	 for (int index = 0;index < text.entities.length;index ++) {

	 TextEntity entity = text.entities[index];
	 TextEntity newEntity = newText.entities[index];

	 if (entity.offset != newEntity.offset) return false;

	 if (entity.length != newEntity.length) return false;

	 if (!entity.type.getClass().equals(newEntity.getClass())) return false;

	 if (entity.type instanceof TextEntityTypePreCode) {

	 TextEntityTypePreCode preCode = (TextEntityTypePreCode)entity.type;
	 TextEntityTypePreCode newCode = (TextEntityTypePreCode)newEntity.type;

	 if (!preCode.language.equals(newCode.language)) return false;

	 } else if (entity.type instanceof TextEntityTypeTextUrl) {

	 TextEntityTypeTextUrl url = (TextEntityTypeTextUrl)entity.type;
	 TextEntityTypeTextUrl newUrl = (TextEntityTypeTextUrl)newEntity.type;

	 if (!url.url.equals(newUrl.url)) return false;

	 } else if (entity.type instanceof TextEntityTypeMentionName) {

	 TextEntityTypeMentionName mention = (TextEntityTypeMentionName)entity.type;
	 TextEntityTypeMentionName newMention = (TextEntityTypeMentionName)newEntity.type;

	 if (mention.userId != newMention.userId) return false;



	 }

	 }

	 return true;

	 }

	 */

}

File: src/main/java/tooko/pm/handlers/EditHandler.kt
/*
 * Copyright (c) 2019 - 2020 KazamaWataru
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package tooko.pm.handlers

import td.TdApi
import td.TdApi.*
import tooko.core.*
import tooko.core.client.*
import tooko.core.raw.*
import tooko.core.utils.*
import tooko.pm.PmBot
import tooko.pm.PmData

class EditHandler : TdBotHandler() {

    override val sudo: PmBot get() = super.sudo as PmBot

    val bot get() = sudo.bot
    val data get() = sudo.data

    override suspend fun onMessageContent(chatId: Long, messageId: Long, newContent: TdApi.MessageContent) {

        val L = bot.owner.langFor

        if (chatId == bot.owner.toLong()) {

            val messageIdStr = messageId.toString() + ""

            if (data.sended.containsKey(messageIdStr)) {

                val targetMessage = data.sended.get(messageIdStr)!!
                val session: PmData.Session = data.getSessionByElement("sended", targetMessage)
                val targetChat: Long = session.chatId

                val oldMessage = getMessageOrNull(targetChat, targetMessage)

                if (oldMessage == null) {

                    data.sended.remove(messageIdStr)
                    session.sended.remove(targetMessage)

                    return

                }

                var edit: TdApi.Function? = null

                if (newContent is MessageText) {

                    edit = sudo make newContent.text to targetChat mkEditAt targetMessage

                } else if (newContent is MessageAnimation || newContent is MessageAudio || newContent is MessageDocument || newContent is MessagePhoto || newContent is MessageVideo) {

                    edit = EditMessageMedia(targetChat, targetMessage, null, newContent.asInput)

                } else if (newContent is MessageVoiceNote) {

                    edit = EditMessageCaption(targetChat, targetMessage, null, newContent.caption)

                } else if (newContent is MessageLocation) {

                    edit = EditMessageLiveLocation(targetChat, targetMessage, null, newContent.location)

                }

                if (edit == null) {

                    sudo make L.EDIT_TYPE_INVALID sendTo chatId

                    return

                }

                try {

                    syncUnit(edit)

                    sudo make L.EDIT_SUCCEED to chatId send deleteDelay()

                } catch (e: TdException) {

                    sudo make L.EDIT_FAILED.input(e.message) sendTo chatId

                }

            }

        } else {

            if (data.received.containsValue(messageId)) {

                val receivedMessageId = data.received.getKey(messageId).toLong()

                val reportMessage = sudo make {

                    inputText = L.EDITED_BY

                    replyToMessageId = receivedMessageId

                } syncTo bot.owner

                val newMessage =sudo make {

                    input = inputForward(chatId, messageId)

                } syncTo bot.owner

                data.sessions.arrayInsert(chatId, "extras", reportMessage.id)
                data.sessions.arrayInsert(chatId, "extras", newMessage.id)

            }

        }

    }

}